### PR TITLE
Provided a better error message when ElementWrapper is passed to a Selector function

### LINKED ISSUE #1366: Provide a better error message when `ElementWrapperList` is passed to a Selector function
Hello there!
Thanks for providing this wonderful library!

I've run in this issue when trying to learn how to write tests with Gauge and Taiko.
I think that a self explainatory message would be helpful to get a better understanding on how to use selector functions.

### Expected behavior

Provide a more detailed error message when a wrong argument is passed to a `Selector` function.

I mean something like:

```
 * My spec  ...[FAIL]
        
        Failed Step: My spec
        Specification: specs/my.spec:8
        Error Message: TypeError: You are passing a `ElementWrapperList` to a `dropDown` selector. Refer https://docs.taiko.dev/api/dropdown/  for the correct parameters
        Stacktrace: 
        TypeError:  You are passing a `ElementWrapperList` to a `dropDown` selector. Refer https://docs.taiko.dev/api/dropdown/  for the correct parameters
            at xpath (node_modules/taiko/lib/helper.js:143:8)
            at node_modules/taiko/lib/taiko.js:2625:53
            at get (node_modules/taiko/lib/elementWrapper/elementWrapperFactory.js:64:56)
            at node_modules/taiko/lib/elementSearch.js:198:28
            at waitUntil (node_modules/taiko/lib/helper.js:121:17)
            at processTicksAndRejections (internal/process/task_queues.js:93:5)
```

### Actual behavior
In a Gauge+Taiko project when `ElementWrapperList` is passed to a Selector function the resulting error is:

```
 * My spec  ...[FAIL]
        
        Failed Step: My spec
        Specification: specs/my.spec:8
        Error Message: TypeError: s.match is not a function
        Stacktrace: 
        TypeError: s.match is not a function
            at xpath (node_modules/taiko/lib/helper.js:143:8)
            at node_modules/taiko/lib/taiko.js:2625:53
            at get (node_modules/taiko/lib/elementWrapper/elementWrapperFactory.js:64:56)
            at node_modules/taiko/lib/elementSearch.js:198:28
            at waitUntil (node_modules/taiko/lib/helper.js:121:17)
            at processTicksAndRejections (internal/process/task_queues.js:93:5)
```

In the Taiko REPL the resulting error is:

```
Error: DropDown with label [object Object]  not found
```

### Steps to reproduce
1. Setup a Gauge + Taiko project
2. Write a `await dropDown($(selector)).select(value);` command inside a Gauge spec.
3. Run the spec

### Versions
```
node v12.13.0, taiko v1.0.13, @getgauge/cli v1.1.1
```

### PR DESCRIPTION
**Pull request by gdorsi**
fixes #1366

### CODE REVIEW

File: test/unit-tests/$.test.js [hunk 1]
```diff
@@ -128,4 +128,12 @@ describe(test_name, () => {
         });
     });
   });
+
+  describe('Parameters validation', () => {
+    it('should throw a TypeError when an ElementWrapper is passad as argument', async () => {
```
Comment (zabil) [inline test/unit-tests/$.test.js:133]:
```suggestion
    it('should throw a TypeError when an ElementWrapper is passed as argument', async () => {
```

File: test/unit-tests/button.test.js [hunk 2]
```diff
@@ -262,4 +262,12 @@ describe(test_name, () => {
       expect(await button('SubmitInLabelDisabled').isDisabled()).to.be.true;
     });
   });
+
+  describe('Parameters validation', () => {
+    it('should throw a TypeError when an ElementWrapper is passad as argument', async () => {
```
Comment (zabil) [inline test/unit-tests/button.test.js:267]:
```suggestion
    it('should throw a TypeError when an ElementWrapper is passed as argument', async () => {
```

File: test/unit-tests/color.test.js [hunk 3]
```diff
@@ -74,4 +74,20 @@ describe('Color picker test', () => {
       'The color code #3233333 is invalid. Please pass a valid HTML color code.',
     );
   });
+
+  describe('Parameters validation', () => {
+    it('should throw a TypeError when an ElementWrapper is passad as argument', async () => {
```
Comment (zabil) [inline test/unit-tests/color.test.js:79]:
```suggestion
    it('should throw a TypeError when an ElementWrapper is passed as argument', async () => {
```

File: test/unit-tests/dropDown.test.js [hunk 4]
```diff
@@ -247,4 +248,12 @@ describe(test_name, () => {
       expect(events).to.eql(['change', 'input']);
     });
   });
+
+  describe('Parameters validation', () => {
+    it('should throw a TypeError when an ElementWrapper is passad as argument', async () => {
```
Comment (zabil) [inline test/unit-tests/dropDown.test.js:253]:
```suggestion
    it('should throw a TypeError when an ElementWrapper is passed as argument', async () => {
```

File: test/unit-tests/image.test.js [hunk 5]
```diff
@@ -159,4 +159,12 @@ describe(test_name, () => {
       await expect(image('#foo').isVisible()).to.be.eventually.rejected;
     });
   });
+
+  describe('Parameters validation', () => {
+    it('should throw a TypeError when an ElementWrapper is passad as argument', async () => {
```
Comment (zabil) [inline test/unit-tests/image.test.js:164]:
```suggestion
    it('should throw a TypeError when an ElementWrapper is passed as argument', async () => {
```

File: test/unit-tests/link.test.js [hunk 6]
```diff
@@ -141,4 +141,12 @@ describe(test_name, () => {
       expect(await thirdElement.text()).to.be.eql('similarLink3');
     });
   });
+
+  describe('Parameters validation', () => {
+    it('should throw a TypeError when an ElementWrapper is passad as argument', async () => {
```
Comment (zabil) [inline test/unit-tests/link.test.js:146]:
```suggestion
    it('should throw a TypeError when an ElementWrapper is passed as argument', async () => {
```

File: test/unit-tests/listItem.test.js [hunk 7]
```diff
@@ -90,4 +90,12 @@ describe(test_name, () => {
       expect(await elems[0].text()).to.be.eql('Coffee');
     });
   });
+
+  describe('Parameters validation', () => {
+    it('should throw a TypeError when an ElementWrapper is passad as argument', async () => {
```
Comment (zabil) [inline test/unit-tests/listItem.test.js:95]:
```suggestion
    it('should throw a TypeError when an ElementWrapper is passed as argument', async () => {
```

File: test/unit-tests/radioButton.test.js [hunk 8]
```diff
@@ -217,4 +218,12 @@ describe(test_name, () => {
       expect(await elements[0].isSelected()).to.be.false;
     });
   });
+
+  describe('Parameters validation', () => {
+    it('should throw a TypeError when an ElementWrapper is passad as argument', async () => {
```
Comment (zabil) [inline test/unit-tests/radioButton.test.js:223]:
```suggestion
    it('should throw a TypeError when an ElementWrapper is passed as argument', async () => {
```

File: test/unit-tests/range.test.js [hunk 9]
```diff
@@ -120,4 +121,12 @@ describe('Range test', () => {
       expect(await range({ id: 'range-2' }).value()).to.be.equal('0');
     });
   });
+
+  describe('Parameters validation', () => {
+    it('should throw a TypeError when an ElementWrapper is passad as argument', async () => {
```
Comment (zabil) [inline test/unit-tests/range.test.js:126]:
```suggestion
    it('should throw a TypeError when an ElementWrapper is passed as argument', async () => {
```

File: test/unit-tests/tablecell.test.js [hunk 10]
```diff
@@ -459,4 +460,12 @@ describe(test_name, () => {
       ).to.be.true;
     });
   });
+
+  describe('Parameters validation', () => {
+    it('should throw a TypeError when an ElementWrapper is passad as argument', async () => {
```
Comment (zabil) [inline test/unit-tests/tablecell.test.js:465]:
```suggestion
    it('should throw a TypeError when an ElementWrapper is passed as argument', async () => {
```

File: test/unit-tests/textBox.test.js [hunk 11]
```diff
@@ -692,4 +693,12 @@ describe(test_name, () => {
       });
     });
   });
+
+  describe('Parameters validation', () => {
+    it('should throw a TypeError when an ElementWrapper is passad as argument', async () => {
```
Comment (zabil) [inline test/unit-tests/textBox.test.js:698]:
```suggestion
    it('should throw a TypeError when an ElementWrapper is passed as argument', async () => {
```

File: test/unit-tests/textMatch.test.js [hunk 12]
```diff
@@ -459,4 +460,31 @@ describe('match', () => {
...
+        retryInterval: 10,
+      });
+    });
+    after(async () => {
+      resetConfig();
+      await closeBrowser();
+      removeFile(filePath);
+    });
+
+    it('should throw a TypeError when an ElementWrapper is passad as argument', async () => {
```
Comment (zabil) [inline test/unit-tests/textMatch.test.js:484]:
```suggestion
    it('should throw a TypeError when an ElementWrapper is passed as argument', async () => {
```

File: test/unit-tests/timeField.test.js [hunk 13]
```diff
@@ -322,4 +323,12 @@ describe(test_name, () => {
       });
     });
   });
+
+  describe('Parameters validation', () => {
+    it('should throw a TypeError when an ElementWrapper is passad as argument', async () => {
```
Comment (zabil) [inline test/unit-tests/timeField.test.js:328]:
```suggestion
    it('should throw a TypeError when an ElementWrapper is passed as argument', async () => {
```

File: test/unit-tests/$.test.js (hunk 1 above)
Comment (gdorsi) [reply test/unit-tests/$.test.js]: Whoops, solved!

**Comment by NivedhaSenthil** [comment]
Instead of doing the check at every selector api, it can be tested at [ElementWrapper's constructor](https://github.com/getgauge/taiko/blob/7cebbea7559be753e6197c35bc44419e2bd1abb0/lib/elementWrapper/elementWrapper.js#L6) so that it checks by default for all selectors. Like 
```
class ElementWrapper {
  constructor(elementType, query, attrValuePairs, _options, ...args) {
    if (attrValuePairs instanceof ElementWrapper) {
      throw new TypeError(
        'You are passing a `ElementWrapperList` to a `' +
          elementType +
          '` selector. Refer https://docs.taiko.dev/api/' +
          elementType.toLowerCase() +
          '/ for the correct parameters',
      );
    }
    const { selector, options } = prepareParameters(attrValuePairs, _options, ...args);
    this.selector = selector;
    this.options = options;
    this._description = desc(selector, query, elementType, options);
  }
```

**Comment by gdorsi** [comment]
> Instead of doing the check at every selector api, it can be tested at [ElementWrapper's constructor](https://github.com/getgauge/taiko/blob/7cebbea7559be753e6197c35bc44419e2bd1abb0/lib/elementWrapper/elementWrapper.js#L6) so that it checks by default for all selectors. Like
> 
> ```
> class ElementWrapper {
>   constructor(elementType, query, attrValuePairs, _options, ...args) {
>     if (attrValuePairs instanceof ElementWrapper) {
>       throw new TypeError(
>         'You are passing a `ElementWrapperList` to a `' +
>           elementType +
>           '` selector. Refer https://docs.taiko.dev/api/' +
>           elementType.toLowerCase() +
>           '/ for the correct parameters',
>       );
>     }
>     const { selector, options } = prepareParameters(attrValuePairs, _options, ...args);
>     this.selector = selector;
>     this.options = options;
>     this._description = desc(selector, query, elementType, options);
>   }
> ```

I tried that but there are some case where `elementType` and the selector name are different ($, tableCell and text).
This made me think that the matches between the selector name and `elementType` were a coincidence.

Anyways we could solve the name mismatches with a simple mapping function.

That would be ok for you?

**Comment by NivedhaSenthil** [comment]
Yes, I feel it would be better if respective child elementWrapper takes that responsibility.

**Comment by NivedhaSenthil** [comment]
Thanks for the contribution @gdorsi

**Pull request merged**